libs/agno/agno/agent/agent.py

```diff
@@ -1033,6 +1030,27 @@ def run(
         self.run_id = run_id
 
         # If no retries are set, use the agent's default retries
+
+        # Parse input into separate variables for processing
+        single_message = None
+        message_list = None
+
+        if input is not None:
+            # Parse the input for processing
+            if isinstance(input, list):
+                # Check if it's a list of Message objects or regular list content
+                if len(input) > 0 and isinstance(input[0], Message):
+                    # List of Message objects - treat as message list
+                    message_list = input
+                elif len(input) > 0 and isinstance(input[0], dict) and "role" in input[0]:
+                    # List of message dicts - convert to Message objects
+                    message_list = [Message.model_validate(m) if isinstance(m, dict) else m for m in input]
+                else:
+                    # Regular list content - treat as single message
+                    single_message = input
+            elif isinstance(input, (Message, dict, str, BaseModel)):
+                # Single content - treat as single message
+                single_message = input
```

Comment:
I feel like this can be combined with the part below? I don't like that we have `single_message` and `message_list` in memory.